### ISSUE TITLE
How to compile the example?

### ISSUE THREAD

**Issue by jabrena** [bmegli/ev3dev-mapping#2]
Good morning @bmegli,

@Burf2000 is trying to run your same LIDAR device but using the tutorial:
http://www.ev3dev.org/docs/tutorials/using-xv11-lidar/

we don't see how to compile this example:
https://github.com/bmegli/ev3dev-mapping/blob/master/experimental/ev3-side/main.cpp

He has uploaded a mini video here:
https://www.youtube.com/watch?v=500PLf-LoOE

but we need your help to compile the example.

Cheers

**Comment by bmegli** [comment]
Hi @jabrena, hi @Burf2000

Guys - I am on holidays ;-)
My wife tells that she will throw the laptop out of the window in a moment.

> we don't see how to compile this example:

There's no how-to to the code. Like it's said in readme.md it's experimental private code that was not meant to be published originally and was unlincensed and published "as is".

I am writing a proper library and this will be supported.in the tutorial.

Yes, we are in touch with @Burf2000.

Anyway, I will upload the whole project (CodeLite) not only single file but I am not sure it will help.
Hopefully tomorrow.

Or better - I will upload the file, and ev3dev bindings used and the command to compile on the brick. 

But still have in mind:
1- The bindings used were from previous kernel release (different device path attribute, jabrena will know what I mean)
2. The code uses gyroscope and 2 motors and laser motor so it will expect all those devices.
3. The code reads from gyroscope using direct attribute. This messes with i2C bus directcly so if you are going to use it, you should know what you're doing. .
4. This code only sends laser and odometry data in UDP pacekts to hardoced ip address. So unless you want to receive the data on PC for further processing you have to modify the code.
5. I am not going to support it in the long run. What can and will be supported is isolated API/library to handle the LIDAR only. Coming soon.

**Comment by bmegli** [comment]
Ok, I have added some additinal information to github repository in readme.md.

If you want to compile it use something along the line:

`g++ main.cpp ev3dev.cpp -DEV3 -O2 -Wall -march=armv5 -std=c++11 -D_GLIBCXX_USE_NANOSLEEP`

This should work both when crosscompiling and from the brick (-march is not necessary in such case). On EV3 you will need at least:

`apt-get install build-essential` to get the build system.

But mind that this is experimental code and is not supported. A proper supported library is on the way.

ev3dev.h and ev3dev.cpp were compatible with kernel I had on the brick. Possibly they will have to be overwrtitten with current versions.

**Comment by jabrena** [comment]
Hi @bmegli,

Many thanks for the notes to compile your program.

@Burf2000 & me are Java developers. The POC is only to test LIDAR.

https://www.youtube.com/watch?v=500PLf-LoOE

I will isolate the code to manage the LIDAR in order to test @Burf2000 at home.

we will inform with the progress.

Cheers

**Comment by burf2000** [comment]
Hey did you ever get any further with this, installing apt-get install build-essential now, I just want to check the sensor works :)

**Comment by bmegli** [comment]
Hello @burf2000

I have added repository with xv11test utility that outputs semicolon separated distances for each angle (as in csv format).

This one is supported and has (hopefully) detailed instructions on building.

Anyway - feel free to ask if there are some problems and we will update the docs together then (say, I am not sure now if gcc is available on fresh ev3dev installation which would be required for make command etc)

**Comment by burf2000** [comment]
Ok so install apt-get install build-essential and then look at the xv11test utility?  thats awesome mate, thank you.  I guess if I could make a socket server in C to send it to a phone, I would be rocking, is it easy to do C on ev3Dev as I have done Objective C for years

**Comment by bmegli** [comment]
@burf2000 Thanks!

And again the same question - did you get it working?

**Comment by bmegli** [comment]
Ok guys, I am closing this.

To sum up:
-we have added `xv11test` utility that can be built with simple `make` command
-we have added `xv11plot.ods` Open Document Spreadsheet that plots the data and explains how to interpret the LIDAR data and apply geometric correction
-as part of `xv11test` we have `xv11lidar.h` and `xv11lidar.c` which can be used as simple C library to read from the LIDAR
- `xv11lidar.c` has a lot of comments now so hopefully can help if one needs more complicated reading scheme (e.g. non-blocking) or wants to read directly from other language

What still could be done in the future:
-motor control for the LIDAR could be integrated in the library with PID so that the LIDAR keeps constant speed regardless of battery voltage
-some examples in other languages might be helpful

@burf2000 , @jabrena  - Thanks for help.
